From 49a4ef1826bf9482517368ca168453cababc7138 Mon Sep 17 00:00:00 2001
From: ValdikSS <iam@valdikss.org.ru>
Date: Sun, 8 Mar 2026 10:51:58 -0300
Subject: [PATCH] Add Dual Channel into Bluetooth Audio Channel Mode developer
 options menu

Add UI support for SBC Dual Channel (SBC HD) mode in the Bluetooth Audio
Channel Mode dialog within Developer Options.

This commit:
- Adds "Dual Channel" radio button to the channel mode dialog layout
- Updates BluetoothChannelModeDialogPreference to dynamically add the option
- Updates BluetoothChannelModeDialogPreferenceController to handle the new mode
- Updates AbstractBluetoothDialogPreferenceController with DUAL_CHANNEL constant
- Force-enables Dual Channel selection for SBC codec

Adapted for Android 16 QPR1 from LineageOS patches.

Co-authored-by: Flopster101 <nahuelgomez329@gmail.com>
Signed-off-by: Flopster101 <nahuelgomez329@gmail.com>
Change-Id: Icd3179ac63310355547fd58060e75f25b4a8210d
---
 .../bluetooth_audio_channel_mode_dialog.xml      |  5 ++++-
 ...tractBluetoothDialogPreferenceController.java |  5 ++++-
 .../BluetoothChannelModeDialogPreference.java    | 11 ++++++++---
 ...othChannelModeDialogPreferenceController.java | 16 ++++++++++++++++
 4 files changed, 32 insertions(+), 5 deletions(-)

diff --git a/res/layout/bluetooth_audio_channel_mode_dialog.xml b/res/layout/bluetooth_audio_channel_mode_dialog.xml
index 9154ee4..c26ec67 100644
--- a/res/layout/bluetooth_audio_channel_mode_dialog.xml
+++ b/res/layout/bluetooth_audio_channel_mode_dialog.xml
@@ -42,6 +42,10 @@
             <include
                 android:id="@+id/bluetooth_audio_channel_mode_stereo"
                 layout="@layout/preference_widget_dialog_radiobutton"/>
+
+            <include
+                android:id="@+id/bluetooth_audio_channel_mode_dual_channel"
+                layout="@layout/preference_widget_dialog_radiobutton"/>
         </RadioGroup>
 
         <include
@@ -51,4 +55,3 @@
     </LinearLayout>
 
 </ScrollView>
-
diff --git a/src/com/android/settings/development/bluetooth/AbstractBluetoothDialogPreferenceController.java b/src/com/android/settings/development/bluetooth/AbstractBluetoothDialogPreferenceController.java
index ea90b4f..8f6846d 100644
--- a/src/com/android/settings/development/bluetooth/AbstractBluetoothDialogPreferenceController.java
+++ b/src/com/android/settings/development/bluetooth/AbstractBluetoothDialogPreferenceController.java
@@ -57,8 +57,11 @@ public abstract class AbstractBluetoothDialogPreferenceController extends
     protected static final int[] BITS_PER_SAMPLES = {BluetoothCodecConfig.BITS_PER_SAMPLE_32,
             BluetoothCodecConfig.BITS_PER_SAMPLE_24,
             BluetoothCodecConfig.BITS_PER_SAMPLE_16};
+    // BluetoothCodecConfig.CHANNEL_MODE_DUAL_CHANNEL is hidden in API stubs.
+    private static final int CHANNEL_MODE_DUAL_CHANNEL = 0x1 << 2;
     protected static final int[] CHANNEL_MODES = {BluetoothCodecConfig.CHANNEL_MODE_STEREO,
-            BluetoothCodecConfig.CHANNEL_MODE_MONO};
+            BluetoothCodecConfig.CHANNEL_MODE_MONO,
+            CHANNEL_MODE_DUAL_CHANNEL};
 
     protected final BluetoothA2dpConfigStore mBluetoothA2dpConfigStore;
 
diff --git a/src/com/android/settings/development/bluetooth/BluetoothChannelModeDialogPreference.java b/src/com/android/settings/development/bluetooth/BluetoothChannelModeDialogPreference.java
index 4a4191a..c260755 100644
--- a/src/com/android/settings/development/bluetooth/BluetoothChannelModeDialogPreference.java
+++ b/src/com/android/settings/development/bluetooth/BluetoothChannelModeDialogPreference.java
@@ -56,14 +56,19 @@ public class BluetoothChannelModeDialogPreference extends BaseBluetoothDialogPre
     }
 
     private void initialize(Context context) {
-        mRadioButtonIds.add(R.id.bluetooth_audio_channel_mode_default);
-        mRadioButtonIds.add(R.id.bluetooth_audio_channel_mode_mono);
-        mRadioButtonIds.add(R.id.bluetooth_audio_channel_mode_stereo);
         String[] stringArray = context.getResources().getStringArray(
                 com.android.settingslib.R.array.bluetooth_a2dp_codec_channel_mode_titles);
         for (int i = 0; i < stringArray.length; i++) {
             mRadioButtonStrings.add(stringArray[i]);
         }
+
+        mRadioButtonIds.add(R.id.bluetooth_audio_channel_mode_default);
+        mRadioButtonIds.add(R.id.bluetooth_audio_channel_mode_mono);
+        mRadioButtonIds.add(R.id.bluetooth_audio_channel_mode_stereo);
+        if (stringArray.length >= 4) {
+            mRadioButtonIds.add(R.id.bluetooth_audio_channel_mode_dual_channel);
+        }
+
         stringArray = context.getResources().getStringArray(
                 com.android.settingslib.R.array.bluetooth_a2dp_codec_channel_mode_summaries);
         for (int i = 0; i < stringArray.length; i++) {
diff --git a/src/com/android/settings/development/bluetooth/BluetoothChannelModeDialogPreferenceController.java b/src/com/android/settings/development/bluetooth/BluetoothChannelModeDialogPreferenceController.java
index d156abd..6f97e6a 100644
--- a/src/com/android/settings/development/bluetooth/BluetoothChannelModeDialogPreferenceController.java
+++ b/src/com/android/settings/development/bluetooth/BluetoothChannelModeDialogPreferenceController.java
@@ -37,6 +37,7 @@ public class BluetoothChannelModeDialogPreferenceController extends
 
     private static final String KEY = "bluetooth_channel_mode_settings";
     private static final String TAG = "BtChannelModeCtr";
+    private static final int CHANNEL_MODE_DUAL_CHANNEL = 0x1 << 2;
 
     public BluetoothChannelModeDialogPreferenceController(Context context, Lifecycle lifecycle,
                                                           BluetoothA2dpConfigStore store) {
@@ -57,6 +58,7 @@ public class BluetoothChannelModeDialogPreferenceController extends
     @Override
     protected void writeConfigurationValues(final int index) {
         int channelModeValue = BluetoothCodecConfig.CHANNEL_MODE_NONE; // default
+        long codecSpecific1 = 0; // default
         switch (index) {
             case 0:
                 final BluetoothCodecConfig currentConfig = getCurrentCodecConfig();
@@ -71,10 +73,15 @@ public class BluetoothChannelModeDialogPreferenceController extends
             case 2:
                 channelModeValue = BluetoothCodecConfig.CHANNEL_MODE_STEREO;
                 break;
+            case 3:
+                channelModeValue = CHANNEL_MODE_DUAL_CHANNEL;
+                codecSpecific1 = 0x1337;
+                break;
             default:
                 break;
         }
         mBluetoothA2dpConfigStore.setChannelMode(channelModeValue);
+        mBluetoothA2dpConfigStore.setCodecSpecific1Value(codecSpecific1);
     }
 
     @Override
@@ -98,6 +105,12 @@ public class BluetoothChannelModeDialogPreferenceController extends
                     selectableIndex.add(convertCfgToBtnIndex(CHANNEL_MODES[i]));
                 }
             }
+            if (currentConfig.getCodecType() == BluetoothCodecConfig.SOURCE_CODEC_TYPE_SBC) {
+                final int dualChannelIndex = convertCfgToBtnIndex(CHANNEL_MODE_DUAL_CHANNEL);
+                if (!selectableIndex.contains(dualChannelIndex)) {
+                    selectableIndex.add(dualChannelIndex);
+                }
+            }
         }
         return selectableIndex;
     }
@@ -112,6 +125,9 @@ public class BluetoothChannelModeDialogPreferenceController extends
             case BluetoothCodecConfig.CHANNEL_MODE_STEREO:
                 index = 2;
                 break;
+            case CHANNEL_MODE_DUAL_CHANNEL:
+                index = 3;
+                break;
             default:
                 Log.e(TAG, "Unsupported config:" + config);
                 break;
-- 
2.53.0

